samza-core/src/main/java/org/apache/samza/storage/SideInputStorageManager.java

```diff
@@ -0,0 +1,372 @@
+ */
+public class SideInputStorageManager {
+  private static final Logger LOG = LoggerFactory.getLogger(SideInputStorageManager.class);
+  private static final String OFFSET_FILE = "SIDE-INPUT-OFFSETS";
+  private static final long STORE_DELETE_RETENTION_MS = TimeUnit.DAYS.toMillis(1); // same as changelog delete retention
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  private final Clock clock;
+  private final Map<String, SideInputProcessor> storeToProcessor;
+  private final Map<String, StorageEngine> stores;
+  private final String storeBaseDir;
+  private final Map<String, Set<SystemStreamPartition>> storeToSSps;
+  private final Map<SystemStreamPartition, String> sspsToStore;
+  private final StreamMetadataCache streamMetadataCache;
+  private final SystemAdmins systemAdmins;
+  private final TaskName taskName;
+  private final JavaStorageConfig storageConfig;
+  private final Map<SystemStreamPartition, String> lastProcessedOffsets = new HashMap<>();
+
+  private Map<SystemStreamPartition, String> startingOffsets;
+
+  public SideInputStorageManager(
+      TaskName taskName,
+      MetricsRegistry metricsRegistry,
+      StreamMetadataCache streamMetadataCache,
+      String storeBaseDir,
+      Map<String, StorageEngine> stores,
+      Map<String, Set<SystemStreamPartition>> storeToSSPs,
+      SystemAdmins systemAdmins,
+      Config config,
+      Clock clock) {
+    this.clock = clock;
+    this.storageConfig = new JavaStorageConfig(config);
+    this.stores = stores;
+    this.storeBaseDir = storeBaseDir;
+    this.storeToSSps = storeToSSPs;
+    this.streamMetadataCache = streamMetadataCache;
+    this.systemAdmins = systemAdmins;
+    this.taskName = taskName;
+
+    validateStoreConfiguration();
+
+    storeToProcessor = stores.keySet().stream()
+        .collect(Collectors.toMap(Function.identity(),
+          storeName -> {
+            SideInputProcessorFactory processorFactory =
+                Util.getObj(storageConfig.getSideInputProcessorFactory(storeName), SideInputProcessorFactory.class);
+            return processorFactory.createInstance(config, metricsRegistry);
+          }));
+
+    this.sspsToStore = new HashMap<>();
+    storeToSSPs.forEach((store, ssps) -> {
+        for (SystemStreamPartition ssp: ssps) {
+          sspsToStore.put(ssp, store);
+        }
+      });
+  }
+
+  /**
+   * Initializes the side input storage manager.
+   */
+  public void init() {
+    LOG.info("Initializing side input stores.");
+    startingOffsets = getStartingOffsets();
+    initializeStoreDirectories();
+  }
+
+  /**
+   * Flushes the contents of the underlying store and writes the offset file to disk.
+   */
+  public void flush() {
+    LOG.info("Flushing the side input stores.");
+    stores.values().forEach(StorageEngine::flush);
+    writeOffsetFiles();
+  }
+
+  /**
+   * Stops the storage engines for all the stores and writes the offset file to disk.
+   */
+  public void stop() {
```

Comment:
Just curious. Shouldn't we call `sideInputManager.stop()` at the end of taskInstance lifecycle? From the implementation, I see it's not called from anywhere.